File: example/InstagramClient/InstagramClient/IGListViewController.m
//
//  IGListViewController.m
//  InstagramClient
//
//  Created by Cristiano Severini on 12/07/12.
//  Copyright (c) 2012 Crino. All rights reserved.
//

#import "IGListViewController.h"

#import "IGAppDelegate.h"

@interface IGListViewController ()

@property(nonatomic, strong) NSArray* data;

@end

@implementation IGListViewController

@synthesize data = _data;

- (id)initWithStyle:(UITableViewStyle)style
{
    self = [super initWithStyle:style];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    
    self.title = @"Followers";
    
    UIBarButtonItem* loginButton = [[UIBarButtonItem alloc] initWithTitle:@"Logout"
                                                                    style:UIBarButtonItemStyleBordered
                                                                   target:self
                                                                   action:@selector(doLogout)];
    self.navigationItem.leftBarButtonItem = loginButton;
    
    IGAppDelegate* appDelegate = (IGAppDelegate*)[UIApplication sharedApplication].delegate;
    NSMutableDictionary* params = [NSMutableDictionary dictionaryWithObjectsAndKeys:@"users/self/followed-by", @"method", nil];
    [appDelegate.instagram requestWithParams:params
                                    delegate:self];
}

- (void)viewDidUnload {
    [super viewDidUnload];
    // Release any retained subviews of the main view.
    // e.g. self.myOutlet = nil;
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // Return the number of sections.
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    // Return the number of rows in the section.
    return [self.data count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    
    // Configure the cell...
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle
                                      reuseIdentifier:CellIdentifier];
    }
    cell.textLabel.text = [[self.data objectAtIndex:indexPath.row] objectForKey:@"username"];
    cell.detailTextLabel.text = [[self.data objectAtIndex:indexPath.row] objectForKey:@"full_name"];
    return cell;
}

/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/

/*
// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath] withRowAnimation:UITableViewRowAnimationFade];
    }   
    else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}
*/

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath
{
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Navigation logic may go here. Create and push another view controller.
    /*
     <#DetailViewController#> *detailViewController = [[<#DetailViewController#> alloc] initWithNibName:@"<#Nib name#>" bundle:nil];
     // ...
     // Pass the selected object to the new view controller.
     [self.navigationController pushViewController:detailViewController animated:YES];
     */
}

-(void)doLogout {
    IGAppDelegate* appDelegate = (IGAppDelegate*)[UIApplication sharedApplication].delegate;
    [appDelegate.instagram logout];
    [self.navigationController popViewControllerAnimated:YES];
}


#pragma mark - IGRequestDelegate

- (void)request:(IGRequest *)request didFailWithError:(NSError *)error {
    NSLog(@"Instagram did fail: %@", error);
    UIAlertView* alertView = [[UIAlertView alloc] initWithTitle:@"Error"
                                                        message:[error localizedDescription]
                                                       delegate:nil
                                              cancelButtonTitle:@"Ok"
                                              otherButtonTitles:nil];
    [alertView show];
}

- (void)request:(IGRequest *)request didLoad:(id)result {
    NSLog(@"Instagram did load: %@", result);
    self.data = (NSArray*)[result objectForKey:@"data"];
    [self.tableView reloadData];
}

@end

File: example/InstagramRubyMotionClient/app/IGListViewController.rb
#
# IGListViewController.rb
# InstagramClient
#
# Created by Devon Blandin on 03/25/13
# http://hello.devonblandin.com
#

class IGListViewController < UITableViewController
  attr_accessor :data

  def viewDidLoad
    super

    self.data = []

    self.view.backgroundColor = UIColor.whiteColor
    self.title = 'Followers'

    logout_button = UIBarButtonItem.alloc.initWithTitle('Logout',
                                                           style: UIBarButtonItemStyleBordered,
                                                          target: self,
                                                          action: 'logout')

    self.navigationItem.leftBarButtonItem = logout_button

    params = { 'method' => 'users/self/followed-by' }
    App.delegate.instagram.requestWithParams(params, delegate: self)
  end

  def viewDidUnload
    super
  end

  def numberOfSectionsInTableView(tableView)
    1
  end

  def tableView(tableView, numberOfRowsInSection: section)
    self.data.count
  end

  def tableView(tableView, cellForRowAtIndexPath: indexPath)
    cell_identifier = 'Cell'

    cell = tableView.dequeueReusableCellWithIdentifier(cell_identifier)
    unless cell
      cell = UITableViewCell.alloc.initWithStyle(UITableViewCellStyleSubtitle, reuseIdentifier: cell_identifier)
    end

    cell.textLabel.text = self.data[indexPath.row]['username']
    cell.detailTextLabel.text = self.data[indexPath.row]['full_name']
    cell
  end

  def logout
    App.delegate.instagram.logout

    self.navigationController.popViewControllerAnimated(true)
  end

  def request(request, didFailWithError: error)
    NSLog("Instagram did fail: #{error}")

    alert_view = UIAlertView.alloc.initWithTitle('Error',
                                                 message: error.localizedDescription,
                                                 delegate: nil,
                                                 cancelButtonTitle: 'Ok',
                                                 otherButtonTitles: nil)

    alert_view.show
  end

  def request(request, didLoad: result)
    NSLog("'Instagram did load: #{result}")

    self.data = result['data']

    self.tableView.reloadData
  end
end
